parsl/executors/flux/executor.py

```diff
@@ -0,0 +1,247 @@
+        try:  # look for the output file
+            with open(expected_file, "rb") as file_handle:
+                future_wrapper.set_result(deserialize(file_handle.read()))
+        except FileNotFoundError:
+            future_wrapper.set_exception(
+                FileNotFoundError(
+                    f"No result found for Parsl task, expected {expected_file}"
+                )
+            )
+        except Exception as unknown_err:
+            future_wrapper.set_exception(unknown_err)
+    else:  # the job exited abnormally
+        future_wrapper.set_exception(
+            AppException(f"Parsl task exited abnormally: returned {returncode}")
+        )
+
+
+class FluxExecutor(NoStatusHandlingExecutor, RepresentationMixin):
+    """Executor that uses Flux to schedule and run jobs.
+
+    Every callable submitted to the executor is wrapped into a Flux job.
+
+    This executor requires that Flux's Python bindings, the ``flux`` package,
+    be available upon startup, and that there be a running Flux instance
+    available.
+
+    Flux jobs are fairly heavyweight. As of Flux v0.25, a single Flux
+    instance is (on many systems) capped at 50 jobs per second. As such,
+    this executor is not a good fit for use-cases consisting of large numbers
+    of small, fast jobs.
+
+    However, Flux is great at handling jobs with large resource requirements,
+    and collections of jobs with varying resource requirements.
+
+    Note that due to vendor-specific extensions, on certain Cray machines like
+    ALCF's Theta or LANL's Trinitite/Trinity, Flux cannot run applications
+    that use the default MPI library. Generally the only workaround is to
+    recompile with another MPI library like OpenMPI.
+
+    Parameters
+    ----------
+    managed : bool
+        If this executor is managed by the DFK or externally handled.
+    working_dir : str
+        Working dir to be used by the executor.
+    label : str
+        Label for this executor instance.
+    flux_handle_args: collections.abc.Sequence
+        Positional arguments to ``flux.Flux()`` instance, if any.
+    flux_handle_kwargs: collections.abc.Mapping
+        Keyword arguments to pass to ``flux.Flux()`` instance, if any.
+        For instance, the ``url`` argument to ``flux.Flux`` can be used to connect to a
+        remote Flux instance at the given URL.
+    """
+
+    def __init__(
+        self,
+        managed: bool = True,
+        working_dir: Optional[str] = "flux_workdir",
+        label: str = "FluxExecutor",
+        flux_handle_args: Sequence = (),
+        flux_handle_kwargs: Mapping = {},
+    ):
+        super().__init__()
+        self.label = label
+        self.working_dir = os.path.abspath(working_dir) if working_dir else os.getcwd()
+        self.managed = managed
+        self._flux_executor = None
+        self.flux_handle_args = flux_handle_args
+        self.flux_handle_kwargs = flux_handle_kwargs
+        self._provider = LocalProvider()
```

Comment:
Flux isn't the default resource manager on any systems yet, so currently the primary way to use Flux is to request an allocation from the local resource manager (like Slurm or LSF) and then launch Flux within that allocation (e.g. srun --ntasks-per-node=1 --mpi=none flux start COMMAND``). How would that work out with the various provides offered by Parsl? I forced ``LocalProvider`` to require that users already have acquired their resource allocation and launched Flux, but that cuts down dramatically on the potential use-cases for this executor.

Reply:
We should probably chat about this. We could essentially have the executor define a flux-specific `launch_cmd` that is then handed off to a parsl provider that will do the instantiation of Flux within a batch job.

Reply:
Right, I was thinking something like that would be a good idea. The actual command will depend on the resource manager, and possibly the platform, though.

For instance, on LLNL Slurm machines we have to add the ``--mpibind=off`` option when we launch Flux in order to get decent performance.

The goal is generally to launch a single instance of Flux on each node in a batch job.

The executor can also connect to a remote Flux instance, but I think LocalProvider would cover that use case.

Reply:
Based on our talk on Tuesday, I updated the implementation so that it properly supports Parsl's execution model and doesn't assume that it is running inside a Flux instance on the target resources. Depending on the setup, users may want to provider overrides to a Launcher instance, or to mess with the executor's ``launch_cmd``. I don't know if there will be any more user-friendly way of doing it, because it is so dependent on the specific platform (LLNL Slurm vs LANL Slurm, for instance).